tests/test_dataclasses.py

```diff
@@ -795,3 +799,32 @@ class Config:
     e.other = 'bulbi2'
     with pytest.raises(dataclasses.FrozenInstanceError):
         e.item.name = 'pika2'
+
+
+def test_pickle_overriden_builtin_dataclass():
```

Comment:
I suggest we add `validate_assignment` in the `Config` to show that the restored dataclass is still a pydantic dataclass with validation. Something like
```python
class Model(BaseModel):
    builtin_dataclass: MyDC

    class Config:
        validate_assignment = True


def test_pickle_overriden_builtin_dataclass():
    m = Model(builtin_dataclass=MyDC(value=3))

    pickled_m = pickle.dumps(m)
    restored_m = pickle.loads(pickled_m)

    assert restored_m.builtin_dataclass.value == 3
    assert restored_m == m

    # ensure the restored dataclass is still a pydantic dataclass
    with pytest.raises(ValidationError):
        m.builtin_dataclass.value = 'qwe'
```

Reply:
@PrettyWood ok

Reply:
@PrettyWood is it actually works like that? I've tried it and failed. I tried changing 
`test_override_builtin_dataclass_nested` in master and also failed:

```


def test_override_builtin_dataclass_nested():
    @dataclasses.dataclass
    class Meta:
        modified_date: Optional[datetime]
        seen_count: int

    @dataclasses.dataclass
    class File:
        filename: str
        meta: Meta

    class Foo(BaseModel):
        file: File

        class Config:
            validate_assignment = True

    FileChecked = pydantic.dataclasses.dataclass(File)
    f = FileChecked(filename=b'thefilename', meta=Meta(modified_date='2020-01-01T00:00', seen_count='7'))
    assert f.filename == 'thefilename'
    assert f.meta.modified_date == datetime(2020, 1, 1, 0, 0)
    assert f.meta.seen_count == 7

    with pytest.raises(ValidationError) as e:
        FileChecked(filename=b'thefilename', meta=Meta(modified_date='2020-01-01T00:00', seen_count=['7']))
    assert e.value.errors() == [
        {'loc': ('meta', 'seen_count'), 'msg': 'value is not a valid integer', 'type': 'type_error.integer'}
    ]

    foo = Foo.parse_obj(
        {
            'file': {
                'filename': b'thefilename',
                'meta': {'modified_date': '2020-01-01T00:00', 'seen_count': '7'},
            },
        }
    )
    assert foo.file.filename == 'thefilename'
    assert foo.file.meta.modified_date == datetime(2020, 1, 1, 0, 0)
    assert foo.file.meta.seen_count == 7

	#
	# New test
	#
	with pytest.raises(ValidationError):
    	foo.file.filename = 'asd'

```

output:
```
Traceback (most recent call last):
  File "/Users/aimestereo/Library/Preferences/PyCharm2019.1/scratches/scratch_82.py", line 17, in <module>
    test_override_builtin_dataclass_nested()
  File "/Users/aimestereo/projects/python/pydantic/tests/test_dataclasses.py", line 715, in test_override_builtin_dataclass_nested
    foo.file.filename = 'asd'
  File "/Users/aimestereo/.pyenv/versions/pydantic-3.7.6/lib/python3.7/site-packages/_pytest/python_api.py", line 728, in __exit__
    fail(self.message)
  File "/Users/aimestereo/.pyenv/versions/pydantic-3.7.6/lib/python3.7/site-packages/_pytest/outcomes.py", line 153, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: DID NOT RAISE <class 'pydantic.error_wrappers.ValidationError'>
```

Reply:
It's normal `Foo.file.filename` is expected to be a string. If you do `foo.file.filename = ['a']` (something that can't be coerced as a string) it should raise an error. And this doesn't for a regular dataclass

Reply:
@PrettyWood oh, I misunderstood what it does. done, with slight test refactor